honeycomb/honeycomb.go

```diff
@@ -122,115 +315,155 @@ func getHoneycombTraceID(traceID string) string {
 	return hcTraceUUID.String()
 }
 
-// Close waits for all in-flight messages to be sent. You should
-// call Close() before app termination.
-func (e *Exporter) Close() {
-	libhoney.Close()
+func honeycombSpan(s *trace.SpanData) *span {
+	sc := s.SpanContext
+
+	hcSpan := &span{
+		TraceID:         getHoneycombTraceID(sc.TraceIDString()),
+		ID:              sc.SpanIDString(),
+		Name:            s.Name,
+		HasRemoteParent: s.HasRemoteParent,
+	}
+	parentID := hex.EncodeToString(s.ParentSpanID[:])
+	var initializedParentID [8]byte
+	if s.ParentSpanID != sc.SpanID && s.ParentSpanID != initializedParentID {
+		hcSpan.ParentID = parentID
+	}
+
+	if s, e := s.StartTime, s.EndTime; !s.IsZero() && !e.IsZero() {
+		hcSpan.DurationMilli = float64(e.Sub(s)) / float64(time.Millisecond)
+	}
+
+	if s.Status != codes.OK {
+		hcSpan.Error = true
+	}
+	return hcSpan
 }
 
-// NewExporter returns an implementation of trace.Exporter that uploads spans to Honeycomb
-//
-// apiKey is your Honeycomb apiKey (also known as your write key)
-// dataset is the name of your Honeycomb dataset
-//
-// Don't have a Honeycomb account? Sign up at https://ui.honeycomb.io/signup
-func NewExporter(config Config) (*Exporter, error) {
+// NewExporter returns an implementation of trace.Exporter that uploads spans to Honeycomb.
+func NewExporter(config Config, opts ...ExporterOption) (*Exporter, error) {
 	// Developer note: bump this with each release
-	versionStr := "0.2.1"
+	// TODO: Stamp this via a variable set at link time with a value derived
+	// from the current VCS tag.
+	const versionStr = "0.2.1"
 	libhoney.UserAgentAddition = "Honeycomb-OpenTelemetry-exporter/" + versionStr
 
-	if config.ApiKey == "" {
-		config.ApiKey = defaultApiKey
+	if len(config.APIKey) == 0 {
+		return nil, errors.New("API key must not be empty")
+	}
+
+	econf := exporterConfig{}
+	for _, o := range opts {
+		if err := o(&econf); err != nil {
+			return nil, err
+		}
 	}
-	if config.Dataset == "" {
-		config.Dataset = defaultDataset
+	if len(econf.dataset) == 0 {
+		econf.dataset = defaultDataset
 	}
 
 	libhoneyConfig := libhoney.Config{
-		WriteKey: config.ApiKey,
-		Dataset:  config.Dataset,
+		WriteKey: config.APIKey,
+		Dataset:  econf.dataset,
 	}
-	if config.Debug {
-		libhoneyConfig.Logger = &libhoney.DefaultLogger{}
+	if len(econf.apiURL) != 0 {
+		libhoneyConfig.APIHost = econf.apiURL
+	}
+	if econf.output != nil {
+		libhoneyConfig.Output = econf.output
 	}
-	if config.APIHost != "" {
```

Comment:
I would like to keep this for internal testing purposes.

Reply:
It's still in there in the `NewExporter` function, consulting `econf.apiURL`.